Add comprehensive memory model tests

reeselevine opened the pull request:
Issue: #963 

- Adds remaining memory model tests from https://gpuharbor.ucsc.edu/webgpu-mem-testing/conformance/
- A total of 40 memory model tests are now included. On my Intel Macbook, the tests finish in ~30 seconds.
- A couple minor updates to the setup code are included so that all tests work correctly and fixed parameters can't be accidentally changed
- Parameter sets for coherence/weak memory tests are updated to use values found to be the most effective when doing experiments

@dneto0 @alan-baker @tyler-utah

P.S. I'm not sure if it's unexpected that `package-lock.json` has changes. If this is bad, I can revert it to the current `package-lock.json` on the main branch.
<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

zoddicus commented on package-lock.json:
```diff
@@ -52,12 +52,13 @@
       }
```
I wouldn't expect this needed to be updated unless you are explicitly using functionality from newer packages, so I would probably revert to main unless there is something specific you need to update for.

zoddicus reviewed (commented):
I have added a review comment about the package-lock.json change, but scanning over the rest of the PR, I don't think I am qualified to approve it in general, since I am not particularly familiar with the memory model and workgroup management in WGSL. Since @dneto0 is away I have added @ben-clayton who hopefully will have enough context to give a proper review.

dneto0 commented:
I plan to review this.  It's in my queue.

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -54,6 +52,22 @@ const storageMemoryCorrTestCode = `
+const corrRMWTestCode = `
+  atomicExchange(&test_locations.value[x_0], 1u);
+  let r0 = atomicLoad(&test_locations.value[x_1]);
+  let r1 = atomicAdd(&test_locations.value[y_1], 0u);
```
So we have two independent dimensions:    storage vs. workgroup; and  read-read vs read-RMW.
So there should be 4 variants, but there are only 3.

Seems we should also test RMW on the workgroup storage.

dneto0 replied:
For consistency with the others, should this be called storageMemoryCorrRMWTestCode ?

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&test_locations.value[y_0], 2u);
+`;
 
+const cowwRMWTestCode = `
```
suggest renaming to storageMemoryCowwRMWTestCode

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&test_locations.value[y_0], 2u);
+`;
+
+const workgroupMemoryCowwTestCode = `
```
Suggest adding a RMW variant too.

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_0].r0, r0);
+`;
+
+const cowrRMWTestCode = `
```
suggest remaing to storageCowrRMWTestCode

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&results.value[id_0].r0, r0);
+`;
+
+const workgroupMemoryCowrTestCode = `
```
Suggest adding a RMW variant

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_0].r0, r0);
+`;
+
+const corw2RMWTestCode = `
```
similarly, rename

dneto0 commented on src/webgpu/shader/execution/memory_model/coherence.spec.ts:
```diff
@@ -98,18 +118,305 @@ g.test('corr')
+  atomicStore(&results.value[id_0].r0, r0);
+`;
+
+const workgroupMemoryCorw2TestCode = `
```
similarly, add a RMW variant for workgroup memory

dneto0 commented on src/webgpu/shader/execution/memory_model/memory_model_setup.ts:
```diff
@@ -959,12 +983,15 @@ export function buildTestShader(
   testType: TestType
 ): string {
   let memoryTypeCode;
+  let isStorageClass = false;
```
FYI.  We renamed "storage class" to "address space", so you can rename this  isStorageAS or something like that.

dneto0 reviewed (changes requested):
Looking really good.
Two requests:
- more consistent naming of the test code snippets.
- add RMW workgroup-storage variants when there are storag-buffer variants.

reeselevine commented:
Added more RMW variants! There are actually three "types" of coherence tests: inter-workgroup on storage address space, intra-workgroup on storage as, and intra-workgroup on workgroup as. I've added rmw variants for all three of these, as well as addressed the other minor comments.

dneto0 reviewed (approved):
I understand even better now.
LGTM, and thanks!
